internal/graph/graph_test.go

```diff
@@ -0,0 +1,346 @@
+package graph
+
+import (
+	"testing"
+
+	"github.com/openfga/openfga/pkg/typesystem"
+	"github.com/stretchr/testify/require"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+)
+
+func TestConnectedObjectGraph_RelationshipIngresss(t *testing.T) {
+
+	tests := []struct {
+		name              string
+		model             *openfgapb.AuthorizationModel
+		targetUserRef     *openfgapb.RelationReference
+		targetObjectRef   *openfgapb.RelationReference
+		expectedIngresses []*RelationshipIngress
+		expectedError     error
+	}{
+		{
+			name: "Direct ingress through ComputedUserset with multiple type restrictions",
```

Comment:
https://dot-to-ascii.ggerganov.com/

input

```
digraph {
    
    "document editor" -> "user"
    "document viewer" -> "document editor"
    "document viewer" -> "user"
}
```

output

```
+-----------------+
| document viewer | -+
+-----------------+  |
  |                  |
  |                  |
  v                  |
+-----------------+  |
| document editor |  |
+-----------------+  |
  |                  |
  |                  |
  v                  |
+-----------------+  |
|      user       | <+
+-----------------+
```

it would be nice to have this as comment for each test

Reply:
dot graphs are cool, and this makes it even easier to use in our document and/or comments in code.

What we really should do is be able to generate a digraph structure from an authorization model.

e.g.

```
type user

type group
  relations
    define member: [user, group#member]

type document
  relations
    define editor: [user, group#member]
    define viewer: [user] as editor
```
then in our graph package we can parse the model and export a D
```
package graph

import (
  ...
  "github.com/emicklei/dot"
)

func ToDigraph(model *openfgapb.AuthorizationModel) (*dot.Graph, error) {
    digraph := dot.NewGraph(dot.Directed)
    // todo: construct the digraph from the model

    return digraph, nil
}
```

Then we can use the Digraph structure and print it out and run it through https://dot-to-ascii.ggerganov.com/ like you did 😄 . That would be awesome 👍 !

Reply:
Tracking this here https://github.com/openfga/openfga/issues/277